src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,466 @@
+package edu.kit.compiler.optimizations.common_subexpression;
+
+import com.sun.jna.Pointer;
+import edu.kit.compiler.optimizations.Optimization;
+import edu.kit.compiler.optimizations.OptimizationState;
+import edu.kit.compiler.optimizations.Util;
+import edu.kit.compiler.optimizations.attributes.Attributes;
+import firm.*;
+import firm.bindings.binding_irdom;
+import firm.bindings.binding_irgopt;
+import firm.bindings.binding_irnode;
+import firm.nodes.*;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+
+import java.util.*;
+
+import static firm.bindings.binding_irgraph.ir_graph_properties_t.IR_GRAPH_PROPERTY_CONSISTENT_DOMINANCE;
+
+@RequiredArgsConstructor
+@Data
+public class CommonSubexpressionElimination implements Optimization.Local {
+
+    // map that holds nodes that are to be exchanged with equivalent other nodes
+    private final HashMap<Node, Node> nodeValues = new HashMap<>();
+    private final Map<Pointer, Map<Pointer, Integer>> dominateMap = new HashMap<>();
+
+    @Override
+    public boolean optimize(Graph g, OptimizationState state) {
+        g.assureProperties(IR_GRAPH_PROPERTY_CONSISTENT_DOMINANCE);
+
+        // compare new change map with old one
+        boolean backEdgesEnabled = BackEdges.enabled(g);
+        if (!backEdgesEnabled) {
+            BackEdges.enable(g);
+        }
+
+        boolean hadChange = false;
+        boolean changes;
+        int maxChanges = 0;
+
+        do {
+            changes = false;
+            nodeValues.clear();
+            dominateMap.clear();
+
+            CSEVisitor visitor = new CSEVisitor(state);
+
+            List<Node> nodes = new ArrayList<>();
+            Util.NodeListFiller nodeCollector = new Util.NodeListFiller(nodes);
+            g.walkTopological(nodeCollector);
+            while (!nodes.isEmpty()) {
+                Node n = nodes.remove(0);
+                n.accept(visitor);
+            }
```

Comment:
Is there a difference to `g.walkTopological(visitor)`?

Reply:
There does seem to be a difference. Probably because `Address` and `Const` nodes are replaces in the visitor. That should probably moved to `transform`.

Reply:
Good point, at first look I thought the Visitor was only for analysis